src/index.js

```diff
@@ -6,83 +6,202 @@ const { ApolloServer, gql } = require('apollo-server');
+    dob: '23/07/1992',
+    interests: ['SciFi', 'Adventure'],
+    isActive: false,
+    favoriteBooks: [1, 2, 3],
+  },
+];
+
+const publications = [
+  {
+    id: 1,
+    name: 'Rolling',
+    author: 'J.K. Rolling',
+    tags: ['SciFi', 'Adventure'],
+  },
+];
+
+const CopyStatus = [
+  {
+    id: 1,
+    name: 'Rolling',
+    author: 'J.K. Rolling',
+    tags: ['SciFi', 'Adventure'],
+    book: 1,
+    borrower: 1,
+    status: 'Checked',
+    quality: 'Excellent',
+  },
+  {
+    id: 2,
+    name: 'Huff',
+    author: 'Loof',
+    tags: ['Adventure'],
+    book: 2,
+    borrower: 1,
+    status: 'Available',
+    quality: 'Good',
+  },
+  {
+    id: 1,
+    name: 'Rolling',
+    author: 'J.K. Rolling',
+    tags: ['SciFi', 'Adventure'],
+    book: 1,
+    borrower: 1,
+    status: 'Checked',
+    quality: 'Excellent',
   },
 ];
 
 // Type definitions define the "shape" of your data and specify
 // which ways the data can be fetched from the GraphQL server.
 const typeDefs = gql`
   # Comments in GraphQL are defined with the hash (#) symbol.
-
   # This "Book" type can be used in other type declarations.
-  type Book {
-    title: String
+
+  interface Publication {
+    id: ID!
+    name: String
     author: User
+    image: String
+    tags: [Tag]
   }
 
-  type User {
+  type Book implements Publication {
+    id: ID!
     name: String
+    author: User
+    image: String
+    tags: [Tag]
+  }
+
+  type CopyStatus implements Publication {
+    id: ID!
+    name: String
+    author: User
+    image: String
+    tags: [Tag]
+    book: Book
+    borrower: User
+    status: BookStatus
+    quality: Quality
+  }
+
+  type User {
+    id: ID!
+    firstName: String
+    lastName: String
     address: String
-    favoriteBook: Book
+    photo: String
+    dob: String
+    interests: [Tag]
+    isActive: Boolean
+    favoriteBooks: [Book]
+  }
+
+  enum Quality {
+    Poor
+    Good
+    Excellent
+  }
+
+  enum BookStatus {
+    Checked
+    Available
+    Missing
+  }
+
+  enum Tag {
+    SciFi
+    Adventure
   }
 
   # The "Query" type is the root of all GraphQL queries.
   # (A "Mutation" type will be covered later on.)
   type Query {
-    books: [Book]
-    users(name: String): [User]
+    books(name: String): [Book]
+    users(firstName: String): [User]
+    tags(tag: String): [Book]
+    interests(interest: String): [User]
   }
 `;
 
 // Resolvers define the technique for fetching the types in the
 // schema.  We'll retrieve books from the "books" array above.
 const resolvers = {
   Query: {
-    books: () => books,
+    books: (obj, args) =>
+      books.filter(book => (args.name ? args.name === book.name : true)),
     users: (obj, args) =>
-      users.filter(user => (args.name ? user.name === args.name : true)),
+      users.filter(
+        user => (args.firstName ? user.firstName === args.firstName : true)
+      ),
+    tags: (obj, args) =>
+      books.filter(book => (args.tag ? book.tags.includes(args.tag) : true)),
```

Comment:
My comment wasn't very clear, my bad. I was thinking we'd add an arg to the `books` query above. So that you can filter the books you get back by tags. you'd chain this onto the `name` filter, like so:

```js
    books: (obj, args) =>
      books.filter(book => (args.name ? args.name === book.name : true))
		   .filter(book => (args.tag ? book.tags.includes(args.tag) : true))
```

Reply:
Lets skip this